File: sms/src/main/java/com/mcw/sms/service/ShipmentService.java
package com.mcw.sms.service;

import com.mcw.sms.model.Event;
import com.mcw.sms.model.Shipment;
import com.mcw.sms.model.State;
import com.mcw.sms.repository.Repository;
import java.util.Optional;

public class ShipmentService {

    private Repository<Shipment> shipmentRepository;
    private Repository<State> stateRepository;
    private Repository<Event> eventRepository;
    private static ShipmentService instance;

    private ShipmentService() {}

    public static synchronized ShipmentService getInstance(Repository<Shipment> shipmentRepository,
                                              Repository<State> stateRepository,
                                              Repository<Event> eventRepository) {
        if (instance == null) {
            instance = new ShipmentService();
            instance.shipmentRepository = shipmentRepository;
            instance.stateRepository = stateRepository;
            instance.eventRepository = eventRepository;
        }
        return instance;
    }

    public void createShipment(String shipmentId, String stateCode) {
        Optional<State> currentState = stateRepository.findOne(stateCode);
        if (currentState.isPresent()) {
            shipmentRepository.save(new Shipment(shipmentId, currentState.get()));
            return;
        }
        throw new IllegalArgumentException("No such state: " + stateCode);
    }

    public Shipment getShipment(String shipmentId) {
        Optional<Shipment> shipment = shipmentRepository.findOne(shipmentId);
        if (shipment.isPresent()) {
            return shipment.get();
        }
        throw new IllegalArgumentException("No such shipment: " + shipmentId);
    }

    public void updateState(StateManager stateManager, String shipmentId, String eventId) {
        Shipment shipment = getShipment(shipmentId);
        Optional<Event> event = eventRepository.findOne(eventId);
        if (event.isPresent()) {
            State nextState = stateManager.getNextState(event.get(), shipment.getCurrentState());
            shipment.addNextState(nextState);
            return;
        }
        throw new IllegalArgumentException("No such event: " + eventId);
    }

}

File: sms_go/service/shipment_service.go
package service

import (
	"context"
	"fmt"
	"sms_go/model"
	"sms_go/repository"
)

type ShipmentService struct {
	shipmentRepository repository.ShipmentRepository
	stateRepository    repository.StateRepository
	eventRepository    repository.EventRepository
	stateManager       *StateManager
}

func NewShipmentService(
	ctx context.Context,
	shipmentRepo repository.ShipmentRepository,
	stateRepository repository.StateRepository,
	eventRepoisotry repository.EventRepository,
	stateTransitionRepository repository.StateTransitionRepository) *ShipmentService {

	stateManager := NewStateManager(ctx, stateTransitionRepository)
	return &ShipmentService{
		shipmentRepository: shipmentRepo,
		stateRepository:    stateRepository,
		eventRepository:    eventRepoisotry,
		stateManager:       stateManager,
	}
}

func (ss *ShipmentService) CreateShipment(ctx context.Context, shipmentId, stateId string) (*model.Shipment, error) {
	currentState := ss.stateRepository.Find(stateId)
	if currentState == nil {
		return nil, fmt.Errorf("no state found with id '%s", stateId)
	}
	shipment := model.NewShipment(shipmentId, currentState)
	ss.shipmentRepository.Save(shipment)
	return shipment, nil
}

func (ss *ShipmentService) UpdateState(ctx context.Context, shipmentId, eventId string) (*model.Shipment, error) {
	shipment := ss.shipmentRepository.Find(shipmentId)
	if shipment == nil {
		return nil, fmt.Errorf("no shipment found with id '%s'", shipmentId)
	}
	event := ss.eventRepository.Find(eventId)
	if event == nil {
		return nil, fmt.Errorf("no event found with id '%s'", eventId)
	}
	nextState, err := ss.stateManager.GetNextState(ctx, shipment.GetCurrentState(), event)
	if err != nil {
		return nil, err
	}
	shipment.AddNextState(nextState)
	return shipment, nil
}
